pkg/operator/ceph/csi/mapping_config.go

```diff
@@ -0,0 +1,258 @@
+
+func (m *CSIDRMappings) addClusterIDMapping(newClusterIDMap map[string]string) {
+	if !m.hasClusterIDMapping(newClusterIDMap) {
+		newDRMap := CSIDRMapping{
+			ClusterIDMapping: newClusterIDMap,
+		}
+		*m = append(*m, newDRMap)
+	}
+}
+
+func (m *CSIDRMappings) addRBDPoolIDMapping(clusterIDMap, newPoolIDMap map[string]string) {
+	for i := 0; i < len(*m); i++ {
+		if reflect.DeepEqual((*m)[i].ClusterIDMapping, clusterIDMap) {
+			(*m)[i].RBDPoolIDMapping = append((*m)[i].RBDPoolIDMapping, newPoolIDMap)
+		}
+	}
+}
+
+func (m *CSIDRMappings) hasClusterIDMapping(newClusterIDMap map[string]string) bool {
+	for _, mapping := range *m {
+		if reflect.DeepEqual(mapping.ClusterIDMapping, newClusterIDMap) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func ReconcileCSIMapConfig(clusterInfo *cephclient.ClusterInfo, ctx *clusterd.Context, pool *cephv1.CephBlockPool, ownerInfo *k8sutil.OwnerInfo) error {
+
+	if pool.Spec.Mirroring.Peers == nil {
+		logger.Infof("no peer secrets added")
+		return nil
+	}
+
+	// Get local Cluster Pool ID
+	localPoolDetails, err := cephclient.GetPoolDetails(ctx, clusterInfo, pool.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get details for the pool %q", pool.Name)
+	}
+
+	mappings := &CSIDRMappings{}
+
+	logger.Debugf("Pool details of local cluster %+v", localPoolDetails)
+
+	for _, peerSecret := range pool.Spec.Mirroring.Peers.SecretNames {
+		s, err := ctx.Clientset.CoreV1().Secrets(clusterInfo.Namespace).Get(context.TODO(), peerSecret, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "failed to fetch kubernetes secret %q bootstrap peer", peerSecret)
+		}
+
+		token := s.Data["token"]
+		decodedTokenToGo, err := decodePeerToken(string(token))
+		if err != nil {
+			return errors.Wrap(err, "failed to decode bootstrap peer token")
+		}
+
+		peerClientName := fmt.Sprintf("client.%s", decodedTokenToGo.ClientID)
+		credentials := cephclient.CephCred{
+			Username: peerClientName,
+			Secret:   decodedTokenToGo.Key,
+		}
+
+		// Add cluster mappings
+		clusterIDMapping := map[string]string{
+			decodedTokenToGo.Namespace: clusterInfo.Namespace,
+		}
+		mappings.addClusterIDMapping(clusterIDMapping)
+
+		// Generate peer cluster keyring in a temporary file
+		keyring := cephclient.CephKeyring(credentials)
+		keyringFile, err := createTmpFile(keyring, "tmp-keyring")
+		if err != nil {
+			return errors.Wrap(err, "failed to create a temp keyring file")
+		}
+		defer removeTmpFile(keyringFile.Name())
+
+		// Generate peer cluster rook config in a temporary file
+		configData := fmt.Sprintf(cfg, decodedTokenToGo.ClusterFSID,
+			decodedTokenToGo.MonHost, peerClientName, keyringFile.Name())
+		configFile, err := createTmpFile(configData, "tmp-config")
+		if err != nil {
+			return errors.Wrap(err, "failed to create a temp config file")
+		}
+		defer removeTmpFile(configFile.Name())
+
+		// Build command
+		args := []string{"osd", "pool", "get", pool.Name, "all",
```

Comment:
This requires a live connection to the peer cluster, right? 
- What if the peer cluster is down because we are in a failure scenario? Hopefully the entry previously detected can remain in the configmap so the mapping would remain in the failure scenario.
- What if we are in the failure scenario already and cannot ever connect to the peer cluster? They would have to update the configmap manually? Or we just require that it was configured before the failure scenario. Expecting pre-configuration seems acceptable, just wanted to be clear on expectations.

Reply:
Yes, this reads the peer cluster token added to the CephBlockPool by the user and establishes connection to the peer cluster. 

Good points. I believe controller reconcile should handle both the scenarios. 

- If the peer cluster is down, controller will fail and reconcile. Previous entries in the config map should remain as `createOrUpdateMappingConfig` won't be called in that case.

- I don't think manually adding the configmap when peer cluster connection has failed would be useful. If the connection can't be established, then even ceph CSI won't be able to reach the peer cluster.  We just log the error and keep reconciling. 

@Madhu-1 thoughts ?

Reply:
Both points make sense to me. It seems reasonable to expect the connection has happened successfully before the other cluster goes down, no need to support manual creation as long as the existing metadata remains even while it's down.

Reply:
Sorry for being late. Both points make sense and we expect this to be created when the pools are created on the cluster. Manual editing of this one as one advantage like if the ceph cluster is recreated with a different pool ID or FSID we can add those entries to this configmap which will help the cephcsi to look at the correct pool or filesystem but this is for the future.